### ISSUE TITLE
Retries: defining fixedBackOff with more than 0 attempt throws KafkaException and does not support DLQ

### ISSUE THREAD

**Issue by darpan7** [spring-projects/spring-kafka#2577]
**In what version(s) of Spring for Apache Kafka are you seeing this issue?**
spring-kafka >> 2.8.11
spring-boot >>  2.7.8

**Describe the bug**

`SeekUtils.seekOrRecover` (SeekUtils.java:208) throws KafkaException if I configure error handler with `fixBackOff(0L, 2L)`. Basically, if I setup max-attempts anything more than 0, it throws an error.

```
if (records == null || !doSeeks(records, consumer, thrownException, true, recovery, container, logger)) { // NOSONAR
	throw new KafkaException("Seek to current after exception", level, thrownException);
}
```
That is because `FailedRecordTracker.recovered(..)` returns false if nextBackOff is anything but -1.
```
		long nextBackOff = failedRecord.getBackOffExecution().nextBackOff();
		if (nextBackOff != BackOffExecution.STOP) {
			if (container == null) {
				Thread.sleep(nextBackOff);
			}
			else {
				ListenerUtils.stoppableSleep(container, nextBackOff);
			}
			return false;
		}
```
**To Reproduce**

Here is my kafka consumer:
```
@Override
    @KafkaListener(
            containerFactory = "batchFactory",
            topics = "test-topic",
            groupId = "localtest-group"
    )
    public void consumeBatch(List<Message<String>> messages, Acknowledgment ack) {
        AtomicInteger i = new AtomicInteger(0);
        messages.forEach(m -> {
            long offset = (long) m.getHeaders().get(KafkaHeaders.OFFSET);
            if(i.get() >= 2){
                log.error(" ******** Failing record {} at offset {}", m.getPayload(), offset);
                throw new BatchListenerFailedException("Failing record at " + offset, i.get());
            }
            i.incrementAndGet();
            log.info("Consuming message [{}] at offset [{}]", m.getPayload(), offset);
        });
       ack.acknowledge();
    }
```

Configuration:
```
@Bean
public ConcurrentKafkaListenerContainerFactory<?, ?> batchFactory() {
        ConcurrentKafkaListenerContainerFactory<Integer, String> factory = new ConcurrentKafkaListenerContainerFactory<>();
        factory.setConsumerFactory(consumerFactory());
        factory.setBatchListener(true);
        factory.errorHandler(errorHandler());
        factory.ackMode(ContainerProperties.AckMode.MANUAL_IMMEDIATE);
        return factory;
}

public ConsumerFactory<Integer, String> consumerFactory() {
      return new DefaultKafkaConsumerFactory<>(consumerConfigs());
}

@Bean
public Map<String, Object> consumerConfigs() {
      // all basic config. Apart from connection details, max.poll.records is set to 5.
}

@Bean
public DefaultErrorHandler errorHandler() {
        DeadLetterPublishingRecoverer recover = new DeadLetterPublishingRecoverer(kafkaTemplate(), (cr, ex) -> {
            return new TopicPartition(cr.topic()+".dead", cr.partition());
        });
        return new DefaultErrorHandler(recover, new FixedBackOff(0L, 2L));
}
```

Stacktrace:

```
org.springframework.kafka.KafkaException: Seek to current after exception; nested exception is org.springframework.kafka.listener.ListenerExecutionFailedException: Listener method 'public void com.local.kafkaClient.consumer.NormalBatchKafkaConsumer.consumeBatch(java.util.List<org.springframework.messaging.Message<java.lang.String>>,org.springframework.kafka.support.Acknowledgment)' threw exception; nested exception is org.springframework.kafka.listener.BatchListenerFailedException: Failing record at 8 @-2; nested exception is org.springframework.kafka.listener.BatchListenerFailedException: Failing record at 8 @-2
	at org.springframework.kafka.listener.SeekUtils.seekOrRecover(SeekUtils.java:208) ~[spring-kafka-2.8.11.jar:2.8.11]
	at org.springframework.kafka.listener.FailedBatchProcessor.seekOrRecover(FailedBatchProcessor.java:195) ~[spring-kafka-2.8.11.jar:2.8.11]
	at org.springframework.kafka.listener.FailedBatchProcessor.doHandle(FailedBatchProcessor.java:136) ~[spring-kafka-2.8.11.jar:2.8.11]
	at org.springframework.kafka.listener.DefaultErrorHandler.handleBatch(DefaultErrorHandler.java:143) ~[spring-kafka-2.8.11.jar:2.8.11]
at org.springframework.kafka.listener.KafkaMessageListenerContainer$ListenerConsumer.invokeBatchErrorHandler(KafkaMessageListenerContainer.java:2325) ~[spring-kafka-2.8.11.jar:2.8.11]
	at org.springframework.kafka.listener.KafkaMessageListenerContainer$ListenerConsumer.doInvokeBatchListener(KafkaMessageListenerContainer.java:2160) ~[spring-kafka-2.8.11.jar:2.8.11]
	at org.springframework.kafka.listener.KafkaMessageListenerContainer$ListenerConsumer.invokeBatchListener(KafkaMessageListenerContainer.java:2024) ~[spring-kafka-2.8.11.jar:2.8.11]
	at org.springframework.kafka.listener.KafkaMessageListenerContainer$ListenerConsumer.invokeListener(KafkaMessageListenerContainer.java:2003) ~[spring-kafka-2.8.11.jar:2.8.11]
	at org.springframework.kafka.listener.KafkaMessageListenerContainer$ListenerConsumer.invokeIfHaveRecords(KafkaMessageListenerContainer.java:1375) ~[spring-kafka-2.8.11.jar:2.8.11]
	at org.springframework.kafka.listener.KafkaMessageListenerContainer$ListenerConsumer.pollAndInvoke(KafkaMessageListenerContainer.java:1366) ~[spring-kafka-2.8.11.jar:2.8.11]
	at org.springframework.kafka.listener.KafkaMessageListenerContainer$ListenerConsumer.run(KafkaMessageListenerContainer.java:1257) ~[spring-kafka-2.8.11.jar:2.8.11]
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) ~[na:na]
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[na:na]
	at java.base/java.lang.Thread.run(Thread.java:834) ~[na:na]
```

**Expected behavior**

I was expecting that record where I am throwing exception should go to dead letter topic after 2 retries. But because of KafkaException, it never reaches to dead-letter topic.

Basically, retries & dlq both are blocked now. Or may be I'm missing something.

**Comment by garyrussell** [comment]
It is not clear what you mean - throwing the exception is intentional - it tells the container that the error was not "handled" and, after the back off (0 in your case) it re-fetches the records starting with the failed one on the next poll. When retries are exhausted, the failed record goes the the DLQ and the seeks are performed on the partitions following the failed record.

**Comment by darpan7** [comment]
Hi Thanks for quick response. even after retries exhausted, it never goes to DLQ. `attemptRecovery()` was never called.

**Comment by garyrussell** [comment]
You are failing a different record each time; that means the first retry is successful and you will never send anything to the DLT. This works as expected...

```java
@SpringBootApplication
public class Kgh2577Application {

	public static void main(String[] args) {
		SpringApplication.run(Kgh2577Application.class, args);
	}

	@KafkaListener(id = "kgh2577", topics = "kgh2577", batch = "true")
	void listen(List<String> in) {
		System.out.println(in);
		for (int i = 0; i < in.size(); i++) {
			if (in.get(i).endsWith("-1")) {
				throw new BatchListenerFailedException("failed", i);
			}
		}
	}

	@Bean
	NewTopic topic() {
		return TopicBuilder.name("kgh2577").partitions(1).replicas(1).build();
	}

	@Bean
	ApplicationRunner runner(KafkaTemplate<String, String> template) {
		return args -> {
			IntStream.range(0, 9).forEach(i -> template.send("kgh2577", "foo-" + i));
		};
	}

	@Bean
	NewTopic topicDlt() {
		return TopicBuilder.name("kgh2577.DLT").partitions(1).replicas(1).build();
	}

	@Bean
	DefaultErrorHandler eh(KafkaTemplate<String, String> template) {
		return new DefaultErrorHandler(new DeadLetterPublishingRecoverer(template), new FixedBackOff(0, 2));
	}


	@KafkaListener(id = "kgh2577.DLT", topics = "kgh2577.DLT")
	void listen(String in) {
		System.out.println("DLT: " + in);
	}

}
```
```properties
spring.kafka.consumer.auto-offset-reset=earliest
spring.kafka.producer.properties.linger.ms=1000
```
```
[foo-0, foo-1, foo-2, foo-3, foo-4, foo-5, foo-6, foo-7, foo-8]
[foo-1, foo-2, foo-3, foo-4, foo-5, foo-6, foo-7, foo-8]
[foo-1, foo-2, foo-3, foo-4, foo-5, foo-6, foo-7, foo-8]
DLT: foo-1
[foo-2, foo-3, foo-4, foo-5, foo-6, foo-7, foo-8]
```

**Comment by darpan7** [comment]
Yup. You're right. Since I'm throwing `BatchListenerFailedException`, it knows at which index it failed. But if I throw any other  exception, it will retry and send the whole batch.
I verified my side and it worked as expected!
Sorry :(

**Comment by Tarun4572** [comment]
@garyrussell so in a batch if a record fails (even after retry) records after that failed record are sent to next poll. am i correct.

**Comment by artembilan** [comment]
That's correct. 
See Javadocs of respective class involved:
```
/**
 * Subclass of {@link FailedRecordProcessor} that can process (and recover) a batch. If
 * the listener throws a {@link BatchListenerFailedException}, the offsets prior to the
 * failed record are committed and the remaining records have seeks performed. When the
 * retries are exhausted, the failed record is sent to the recoverer instead of being
 * included in the seeks. If other exceptions are thrown, the fallback handler takes the processing.
 *
 * @author Gary Russell
 * @author Francois Rosiere
 * @author Wang Zhiyang
 * @author Artem Bilan
 * @since 2.8
 *
 */
public abstract class FailedBatchProcessor extends FailedRecordProcessor {
```
